File: app/src/main/java/com/inteligenciadigital/netflixremake/MovieActivity.java
package com.inteligenciadigital.netflixremake;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.inteligenciadigital.netflixremake.model.Movie;
import com.inteligenciadigital.netflixremake.model.MovieDetail;
import com.inteligenciadigital.netflixremake.util.ImageDownloadTask;
import com.inteligenciadigital.netflixremake.util.MovieDetailTask;

import java.util.ArrayList;
import java.util.List;

public class MovieActivity extends AppCompatActivity implements MovieDetailTask.MovieDetailLoader {

	private TextView txtTitle;
	private TextView txtDesc;
	private TextView txtCast;
	private ImageView imgCover;
	private MovieAdapter movieAdapter;
	private RecyclerView recyclerView;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_movie);

		this.txtTitle = findViewById(R.id.text_view_title);
		this.txtDesc = findViewById(R.id.text_view_desc);
		this.txtCast = findViewById(R.id.text_view_cast);
		this.imgCover = findViewById(R.id.image_view_cover);
		this.recyclerView = findViewById(R.id.recycler_view_similar);

		Toolbar toolbar = findViewById(R.id.toolbar);
		this.setSupportActionBar(toolbar);

		if (this.getSupportActionBar() != null) {
			this.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
			this.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_baseline_arrow_back_24);
			this.getSupportActionBar().setTitle(null);
		}

		List<Movie> movies = new ArrayList<>();
		this.movieAdapter = new MovieAdapter(movies);
		this.recyclerView.setAdapter(this.movieAdapter);
		this.recyclerView.setLayoutManager(new GridLayoutManager(this, 3));

		Bundle extras = this.getIntent().getExtras();
		if (extras != null) {
			int id = extras.getInt("id");
			MovieDetailTask movieDetailTask = new MovieDetailTask(this);
			movieDetailTask.setMovieDetailLoader(this);
			movieDetailTask.execute("https://tiagoaguiar.co/api/netflix/" + id);
		}
	}

	@Override
	public boolean onOptionsItemSelected(@NonNull MenuItem item) {
		if (item.getItemId() == android.R.id.home) this.finish();

		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onResult(MovieDetail movieDetail) {
		this.txtTitle.setText(movieDetail.getMovie().getTitle());
		this.txtDesc.setText(movieDetail.getMovie().getDesc());
		this.txtCast.setText(movieDetail.getMovie().getCast());

		ImageDownloadTask imageDownloadTask = new ImageDownloadTask(this.imgCover);
		imageDownloadTask.setShadowEnable(true);
		imageDownloadTask.execute(movieDetail.getMovie().getCoverUrl());

		this.movieAdapter.setMovies(movieDetail.getMoviesSimilar());
		this.movieAdapter.notifyDataSetChanged();
	}

	private static class MovieHolder extends RecyclerView.ViewHolder {

		private final ImageView imageViewCover;

		public MovieHolder(@NonNull View itemView) {
			super(itemView);
			this.imageViewCover = itemView.findViewById(R.id.image_view_cover);
		}
	}

	private class MovieAdapter extends RecyclerView.Adapter<MovieActivity.MovieHolder> {

		private List<Movie> movies;

		public MovieAdapter(List<Movie> movies) {
			this.movies = movies;
		}

		@NonNull
		@Override
		public MovieActivity.MovieHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			View view = getLayoutInflater()
					.inflate(R.layout.movie_item_similar, parent, false);
			return new MovieHolder(view);
		}

		@Override
		public void onBindViewHolder(@NonNull MovieActivity.MovieHolder holder, int position) {
			Movie movie = this.movies.get(position);
			new ImageDownloadTask(holder.imageViewCover).execute(movie.getCoverUrl());
		}

		@Override
		public int getItemCount() {
			return this.movies.size();
		}

		public void setMovies(List<Movie> movies) {
			this.movies.clear();
			this.movies = movies;
		}
	}
}

File: app/src/main/java/com/inteligenciadigital/netflixremake/kotlin/MovieActivity.kt
package com.inteligenciadigital.netflixremake.kotlin

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.inteligenciadigital.netflixremake.R
import com.inteligenciadigital.netflixremake.model.Movie
import com.inteligenciadigital.netflixremake.util.ImageDownloadTask
import com.inteligenciadigital.netflixremake.util.MovieDetailTask
import kotlinx.android.synthetic.main.activity_movie.*
import kotlinx.android.synthetic.main.activity_movie.image_view_cover
import kotlinx.android.synthetic.main.movie_item_similar.view.*

class MovieActivity : AppCompatActivity() {

	private lateinit var movieAdapter: MovieAdapter

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		this.setContentView(R.layout.activity_movie)

		intent.extras?.let {
			val id = it.getInt("id")
			val task = MovieDetailTask(this)
			task.setMovieDetailLoader { movieDetail ->
				text_view_title.text = movieDetail.movie.title
				text_view_desc.text = movieDetail.movie.desc
				text_view_cast.text = getString(R.string.cast, movieDetail.movie.cast)

				ImageDownloadTask(image_view_cover).apply {
					this.setShadowEnable(true)
					this.execute(movieDetail.movie.coverUrl)
				}

				this.movieAdapter.movies.clear()
				this.movieAdapter.movies.addAll(movieDetail.moviesSimilar)
				this.movieAdapter.notifyDataSetChanged()
			}

			task.execute("https://tiagoaguiar.co/api/netflix/$id")

			this.setSupportActionBar(toolbar)

			supportActionBar?.let { toolbar ->
				toolbar.setDisplayHomeAsUpEnabled(true)
				toolbar.setHomeAsUpIndicator(R.drawable.ic_baseline_arrow_back_24)
				toolbar.title = null
			}

			val movies = arrayListOf<Movie>()
			this.movieAdapter = MovieAdapter(movies)
			recycler_view_similar.adapter = this.movieAdapter
			recycler_view_similar.layoutManager = GridLayoutManager(this, 3)
		}
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		when(item.itemId) {
			android.R.id.home -> this.finish()
		}
		return super.onOptionsItemSelected(item)
	}

	private inner class MovieAdapter(val movies: MutableList<Movie>) : RecyclerView.Adapter<MovieHolder>() {
		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieHolder = MovieHolder(
				layoutInflater.inflate(R.layout.movie_item_similar, parent, false)
		)

		override fun onBindViewHolder(holder: MovieHolder, position: Int) = holder.bind(movies[position])

		override fun getItemCount(): Int = movies.size
	}

	private class MovieHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
		fun bind(movie: Movie) {
			with(itemView) {
				ImageDownloadTask(image_view_cover).execute(movie.coverUrl)
			}
		}
	}
}